proof/refine/ARM/Finalise_R.thy

```diff
@@ -3116,21 +3295,153 @@ crunches cancelAllIPC, cancelAllSignals, unbindNotification, unbindMaybeNotifica
 
+lemma setQueue_after_removeFromBitmap:
+  "(setQueue d p q >>= (\<lambda>rv. (when P (removeFromBitmap d p)) >>= (\<lambda>rv. threadSet f t))) =
+   (when P (removeFromBitmap d p) >>= (\<lambda>rv. (threadSet f t) >>= (\<lambda>rv. setQueue d p q)))"
+  supply bind_assoc[simp add]
+  apply (case_tac P, simp_all)
+   prefer 2
+   apply (simp add: setQueue_after)
+  apply (simp add: setQueue_def when_def)
+  apply (subst oblivious_modify_swap)
+   apply (fastforce simp: threadSet_def getObject_def setObject_def
+                          loadObject_default_def bitmap_fun_defs
+                          split_def projectKO_def2 alignCheck_assert
+                          magnitudeCheck_assert updateObject_default_def
+                   intro: oblivious_bind)
+  apply clarsimp
+  done
+
+lemma valid_inQ_queues_exceptI[intro]:
+  "valid_inQ_queues s \<Longrightarrow> valid_inQ_queues_except t s"
+  by (simp add: valid_inQ_queues_except_def valid_inQ_queues_def)
+
+lemma threadSet_valid_inQ_queues_dequeue_wp:
+ "\<lbrace>valid_inQ_queues_except t and (\<lambda>s. \<forall>d p. t \<notin> set (ksReadyQueues s (d,p)))\<rbrace>
+  threadSet (tcbQueued_update (\<lambda>_. False)) t
+  \<lbrace>\<lambda>_. valid_inQ_queues \<rbrace>"
+  unfolding threadSet_def
+  apply (rule hoare_seq_ext[OF _ getObject_tcb_sp])
+  apply (simp add: valid_inQ_queues_def)
+  apply (wpsimp wp: hoare_Ball_helper hoare_vcg_all_lift setObject_tcb_strongest)
+  apply (clarsimp simp: valid_inQ_queues_except_def)
+  done
+
+lemma removeFromBitmap_valid_inQ_queues_except[wp]:
+  "removeFromBitmap d p \<lbrace>valid_inQ_queues_except t\<rbrace>"
+  unfolding bitmapQ_defs valid_inQ_queues_except_def
+  by (wpsimp simp: bitmap_fun_defs)+
+
+lemma setQueue_valid_inQ_queues_except_dequeue_wp:
+  "\<And>d p ts t.
+   \<lbrace>\<lambda>s. valid_inQ_queues_except t s \<and> (\<forall>t' \<in> set ts. obj_at' (inQ d p) t' s)
+        \<and> t \<notin> set ts \<and> distinct ts\<rbrace>
+   setQueue d p ts
+   \<lbrace>\<lambda>_. valid_inQ_queues_except t \<rbrace>"
+  unfolding setQueue_def valid_inQ_queues_except_def null_def
+  by wp force
+
+lemma valid_queues_no_bitmap_correct_queueI[intro]:
+  "valid_inQ_queues s \<Longrightarrow> correct_queue t s"
+  by (fastforce simp: correct_queue_def valid_inQ_queues_def obj_at'_def inQ_def)
+
+lemma tcbSchedDequeue_valid_inQ_queues_weak:
+  "\<lbrace>valid_inQ_queues_except t and correct_queue t and tcb_at' t\<rbrace>
+   tcbSchedDequeue t
+   \<lbrace>\<lambda>_. valid_inQ_queues\<rbrace>"
+  unfolding tcbSchedDequeue_def null_def valid_inQ_queues_def
+  apply wp (* stops on threadSet *)
+          apply (rule hoare_post_eq[OF _ threadSet_valid_inQ_queues_dequeue_wp]
+                 , simp add: valid_inQ_queues_def)
+         apply (wp hoare_vcg_if_lift)+
+           apply (wp hoare_vcg_imp_lift setQueue_valid_inQ_queues_except_dequeue_wp
+                     threadGet_const_tcb_at)+
+  (* wp done *)
+  apply (normalise_obj_at')
```

Comment:
nitpick: no parentheses needed here, right? The same for two lines below

Reply:
Yep, you're right. Removed